modules/fedimint-mint-client/src/redemption.rs

```diff
@@ -0,0 +1,279 @@
+use fedimint_client::sm::{ClientSMDatabaseTransaction, OperationId, State, StateTransition};
+use fedimint_client::transaction::{ClientInput, TransactionBuilder};
+use fedimint_client::DynGlobalClientContext;
+use fedimint_core::core::DynInput;
+use fedimint_core::encoding::{Decodable, Encodable};
+use fedimint_core::{TieredMulti, TransactionId};
+use fedimint_mint_common::MintInput;
+
+use crate::{MintClientContext, SpendableNote};
+
+// TODO: add retry with valid subset of e-cash notes
+/// State machine managing the e-cash redemption process related to a mint
+/// input.
+///
+/// ```mermaid
+/// graph LR
+///     classDef virtual fill:#fff,stroke-dasharray: 5 5
+///
+///     Created -- containing tx accepted --> Success
+///     Created -- containing tx rejected --> Refund
+///     Refund -- refund tx rejected --> Error
+///     Refund -- refund tx accepted --> RS[Refund Success]
+/// ```
+#[derive(Debug, Clone, Eq, PartialEq, Decodable, Encodable)]
+pub enum MintRedemptionStates {
+    Created(MintRedemptionStateCreated),
+    Refund(MintRedemptionStateRefund),
+    Success(MintRedemptionStateSuccess),
+    Error(MintRedemptionStateError),
+    RefundSuccess(MintRedemptionStateRefundSuccess),
+}
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Decodable, Encodable)]
+pub struct MintRedemptionCommon {
+    pub(crate) operation_id: OperationId,
+    pub(crate) txid: TransactionId,
+    pub(crate) input_idx: u64,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Decodable, Encodable)]
+pub struct MintRedemptionStateMachine {
+    pub(crate) common: MintRedemptionCommon,
+    pub(crate) state: MintRedemptionStates,
+}
+
+impl State for MintRedemptionStateMachine {
+    type ModuleContext = MintClientContext;
+    type GlobalContext = DynGlobalClientContext;
+
+    fn transitions(
+        &self,
+        context: &Self::ModuleContext,
```

Comment:
Could we somehow aggregate these two types & values into one?

Reply:
Probably, didn't seem like a huge RoI thing yet.